File: src/advent_of_code_2015/Day12/main.cs
using System;
using System.IO;
using System.Text.Json;

class Program
{
    static int Sum(JsonElement element)
    {
        int total = 0;

        switch (element.ValueKind)
        {
            case JsonValueKind.Number:
                total += element.GetInt32();
                break;

            case JsonValueKind.Array:
                foreach (var item in element.EnumerateArray())
                    total += Sum(item);
                break;

            case JsonValueKind.Object:
                foreach (var prop in element.EnumerateObject())
                {
                    if (prop.Value.ValueKind == JsonValueKind.String && prop.Value.GetString() == "red")
                        return 0;
                    else 
                        total += Sum(prop.Value);
                }
                   
                break;
        }

        return total;
    }

    static void Main()
    {
        string json = File.ReadAllText("input.txt");

        JsonDocument doc = JsonDocument.Parse(json);

        Console.WriteLine(Sum(doc.RootElement));
    }
}

File: src/advent_of_code_2015/Day12/main.py
import json


def add_all_numbers(obj) -> int:
    if isinstance(obj, int):
        return obj

    if isinstance(obj, list):
        return sum(add_all_numbers(x) for x in obj)

    if isinstance(obj, dict):
        if "red" in obj.values():
            return 0
        
        return sum(add_all_numbers(v) for v in obj.values())

    return 0

with open("input.txt") as file:
    data = json.load(file)

print(add_all_numbers(data))

